codec/decode.go

```diff
@@ -0,0 +1,111 @@
+package codec
+
+import (
+	"encoding/binary"
+	"errors"
+)
+
+// DecodeInteger accepts a byte array representing a SCALE encoded integer and performs SCALE decoding of the
+// int, then returns it
+func DecodeInteger(b []byte) (int64, error) {
+	if len(b) == 1 {
+		return int64(b[0] >> 2), nil
+	} else if len(b) == 2 {
+		o := binary.LittleEndian.Uint16(b) >> 2
+		return int64(o), nil
+	} else if len(b) == 4 {
+		o := binary.LittleEndian.Uint32(b) >> 2
+		return int64(o), nil
+	}
+
+	var o int64
+	var err error
+
+	topSixBits := (binary.LittleEndian.Uint16(b) & 0xff) >> 2
+	byteLen := topSixBits + 4
+
+	if byteLen == 4 {
+		o = int64(binary.LittleEndian.Uint32(b[1 : byteLen+1]))
+	} else if byteLen > 4 && byteLen < 8 {
+		upperBytes := make([]byte, 8-byteLen)
+		upperBytes = append(b[5:byteLen+1], upperBytes...)
+		upper := int64(binary.LittleEndian.Uint32(upperBytes)) << 32
+		lower := int64(binary.LittleEndian.Uint32(b[1:5]))
+		o = upper + lower
+	}
+
+	if o == 0 {
```

Comment:
How do you know we didn't just decode `0`?

Reply:
if we decoded 0 it should have happened in the first case (1 byte mode), it is invalid if it happens in the last case. so I think it's ok to throw an error here, as the bytes in a large integer >2**30 shouldn't be zero